src/lib/y2storage/autoinst_proposal.rb

```diff
@@ -0,0 +1,184 @@
+    # @param drives      [Proposal::AutoinstDrivesMap] Devices map from an AutoYaST profile
+    # @return [Devicegraph] Devicegraph containing the planned devices
+    def propose_devicegraph(devicegraph, drives)
+      planned_devices = plan_devices(devicegraph, drives)
+      create_devices(devicegraph, planned_devices, drives.disk_names)
+    rescue Error
+      log.info "No partitions were specified. Falling back to guided setup planning."
+      propose_guided_devicegraph(devicegraph, drives)
+    end
+
+    # Finds a suitable devicegraph using the guided proposal approach
+    #
+    # If the :desired target fails, it will retry with the :min target.
+    #
+    # @param devicegraph [Devicegraph]       Starting point
+    # @param drives      [AutoinstDrivesMap] Devices map from an AutoYaST profile
+    # @return [Devicegraph] Proposed devicegraph using the guidede proposal approach
+    #
+    # @raise [Error] No suitable devicegraph was found
+    # @see proposed_guided_devicegraph
+    def propose_guided_devicegraph(devicegraph, drives)
+      guided_devicegraph_for_target(devicegraph, drives, :desired)
+    rescue Error
+      guided_devicegraph_for_target(devicegraph, drives, :min)
+    end
+
+    # Calculates list of planned devices
+    #
+    # If the list does not contain any partition, then it follows the same
+    # approach than the guided proposal
+    #
+    # @param devicegraph [Devicegraph]                 Starting point
+    # @param drives      [Proposal::AutoinstDrivesMap] Devices map from an AutoYaST profile
+    # @return [Array<Planned::Device>] Devices to add
+    def plan_devices(devicegraph, drives)
+      planner = Proposal::AutoinstDevicesPlanner.new(devicegraph)
+      planner.planned_devices(drives)
+    end
+
+    # Creates a devicegraph using the same approach than guided partitioning
+    #
+    # @param devicegraph [Devicegraph]       Starting point
+    # @param drives      [AutoinstDrivesMap] Devices map from an AutoYaST profile
+    # @param target      [Symbol] :desired means the sizes of the partitions should
+    #   be the ideal ones, :min for generating the smallest functional partitions
+    def guided_devicegraph_for_target(devicegraph, drives, target)
+      guided_settings = proposal_settings_for_disks(drives.disk_names)
+      guided_planner = Proposal::PlannedDevicesGenerator.new(guided_settings, devicegraph)
+      planned_devices = guided_planner.planned_devices(target)
+      create_devices(devicegraph, planned_devices, drives.disk_names)
+    end
+
+    # Creates planned devices on a given devicegraph
+    #
+    # @param devicegraph     [Devicegraph]            Starting point
+    # @param planned_devices [Array<Planned::Device>] Devices to add
+    # @return [Devicegraph]
+    def create_devices(devicegraph, planned_devices, disk_names)
+      devices_creator = Proposal::AutoinstDevicesCreator.new(devicegraph)
+      devices_creator.devicegraph(planned_devices, disk_names)
```

Comment:
ok, so it can create different devicegraph using constructor argument as initial one. I think it really helps to mention that it is copy of initial one in constructor. For me this is same case as String and its method. To mention that it returns copy, as graph is in fact not! argument, but part of object.

About naming. If I try to follow "say, not ask" I prefer names like `devicegraph_for` or `create` as it say what action we want. Simple `devicegraph` is for me like asking for devicegraph.